### PR TITLE
Add SAM models of MSRE insertion tests

### PR DESCRIPTION
**Pull request by hapfang**
This PR contains the SAM models and documentation for the system models of MSRE reactivity insertion tests at the power level of 5 MW. The Point Kinetic Equation (PKE) model is used in the SAM model with two reactivity feedback mechanisms: the fuel density reactivity feedback, and the moderator temperature feedback. The SAM models have been tested right before this PR on ANL cluster, and the models produce expected results.

### CODE REVIEW

File: msr/msre/reactivity_insertion/msre_insertion.in [hunk 1]
```diff
@@ -0,0 +1,440 @@
+# MSRE Model for reactivity insertion test
```
Comment (GiudGiud) [inline msr/msre/reactivity_insertion/msre_insertion.in:1]:
```suggestion
# MSRE Model for reactivity insertion test
# SAM input file for transient simulation
```

File: doc/content/msr/msre/msre_sam_model.md [hunk 2]
```diff
@@ -147,3 +147,15 @@ The U-tube primiary heat exchanger cools down the fuel salt, which returns to th
...
...as achieved, i.e., all parameters (density, temperature, etc.) in the system no longer change. Some preliminary scoping studies are performed to ensure the initial conditions (core inlet/outlet temperature, HX secondary side inlet/outlet temperature) are consistent with the experiments before this step. In the second step, an external reactivity is inserted at the beginning of the simulation, and power is allowed to change due to the external reactivity insertion and the reactivity feedback following the power increase. This MSRE SAM model is similar to the existing SAM model, except that the MSR-specific point kinetics is enabled and the MSRE primary heat exchanger is modeled by a `PBHeatExchanger` component in SAM. The Point Kinetic Equation for MSRs is employed to predict the power evolution. The delayed neutron precursor sources are specified in the core, upper and lower plenum. The power fractions in these regions are 87.5%, 3.89%, and 8.59%, respectively. The delayed neutron precursor sources are directly proportional to the power fraction in the current code implementation. Thus, to account for all delayed neutron precursors, no power is specified in the graphite moderators.
```
Comment (GiudGiud) [inline doc/content/msr/msre/msre_sam_model.md:154]:
```suggestion
An additional SAM model was developed for the MSRE reactivity insertion tests. There are three external reactivity insertion experiments conducted in MSRE at power levels of 1 MW, 5 MW, and 8 MW. This SAM model is developed for the 5 MW case. The SAM model consists of two calculation steps. In Step 1, the power level is adjusted to be consistent with that in the experiments. The power is kept constant for a long period of time (1000 - 2000 seconds) until the steady state condition was achieved, i.e. all parameters (density, temperature, etc.) in the system no longer change. Some preliminary scoping studies are performed to ensure the initial conditions (core inlet/outlet temperature, HX secondary side inlet/outlet temperature) are consistent with the experiments before this step. In the second step, an external reactivity is inserted at the beginning of the simulation, and power is allowed to change due to the external reactivity insertion and the reactivity feedback following the power increase. This MSRE SAM model is similar to the existing SAM model, except that the MSR-specific point kinetics is enabled and the MSRE primary heat exchanger is modeled by a `PBHeatExchanger` component in SAM. The Point Kinetic Equation for MSRs is employed to predict the power evolution. The delayed neutron precursor sources are specified in the core, upper and lower plenum. The power fractions in these regions are 87.5%, 3.89%, and 8.59%, respectively. The delayed neutron precursor sources are directly proportional to the power fraction in the current code implementation. Thus, to account for all delayed neutron precursors, no power is specified in the graphite moderators.
```

File: msr/msre/reactivity_insertion/msre_pke_ss.in [hunk 3]
```diff
@@ -0,0 +1,444 @@
+# MSRE Model for reactivity insertion test
```
Comment (GiudGiud) [inline msr/msre/reactivity_insertion/msre_pke_ss.in:1]:
```suggestion
# MSRE Model for reactivity insertion test
# SAM input file for steady state initialization
```

corrct me if I m wrong I havent run them yet

**Review by GiudGiud — CHANGES_REQUESTED** [review]
what is this .in extension?
If you do this instead of .i, I think you will loose support from the plugins in the text editors

I can add the testing

Could you please add your model to some of the general indexing of models? As appropriate
I m thinking mostly this one
https://mooseframework.inl.gov/vtb/resources/simulation_type.html

**Comment by hapfang** [comment]
Thanks @GiudGiud. The *.in is the same as *.i, but I should have changed them just to be consistent. I can put the model tentatively under "Reactivity insertion accident (RIA)", but can I do that for a subsection from the general MSRE SAM model page?

**Comment by GiudGiud** [comment]
You can use a more accurate link than the ones i ususally use, by linking to a subsection not just a page. 
Like this for example
```
[AdvancedOutput](syntax/Outputs/index.md#advanced-output).
```

btw could you please add an issue reference to the commits, that way the test suite runs at least even without the new test

**Comment by hapfang** [comment]
Hi @GiudGiud, I have made the changes as suggested. The MSRE reactivity insertion tests model is listed under the section RIA with a link to the specific documentation section on the MSRE page.

**Review by hapfang — COMMENTED** [review]
the changes look good to me.

**Comment by hapfang** [comment]
Hi @GiudGiud, what else do I need to change for this PR? It has been waiting for 10 days. Thanks!

**Comment by GiudGiud** [comment]
Hi Jun

it s still missing testing. I need to get to it to add it basically but have high priority items on the moose side to deal with first

Guillaume

**Comment by hapfang** [comment]
OK, thanks Guillaume. Please prioritize your own tasks. 
Let me know what I can help with this PR when you have time to take care of it.

**Comment by GiudGiud** [comment]
@hapfang could you please give me write access to your branch on [hapfang/virtual_test_bed](https://github.com/hapfang/virtual_test_bed) ? Got testing added and a fix or two but cant push them rn, getting:
```
error: Authentication error: Authentication required: You must have push access to verify locks
error: failed to push some refs to 'github.com:hapfang/virtual_test_bed.git'
```

**Comment by hapfang** [comment]
Thanks @GiudGiud for looking into this PR, I have added you as a collaborator to my forked repo.

**Comment by GiudGiud** [comment]
@hapfang I m having trouble getting reproducible results on this input
https://civet.inl.gov/job/1147382/
there's a lot of diffs on a lot of outputs, with tighter tolerances. Any chance you could have a look at it?

**Comment by hapfang** [comment]
Hi @GiudGiud, I am not familiar with civet. What are the test and reference data based upon in terms of the SAM versions? I will see if I can reproduce the issue. Thanks.

**Comment by GiudGiud** [comment]
ok I ll get them both to be the same then I ll let you know

**Comment by hapfang** [comment]
It is concerning that the 1-D model results show a dependency on SAM versions. if possible, I still would like to dig deeper to identify the causes though.

**Comment by GiudGiud** [comment]
I used this version SAM Version: derived from v0.9.5.1, git commit 4f2b2440 on 2022-08-03
it should be consistent with the only one now

**Comment by GiudGiud** [comment]
No objection from me and I m happy to work on it. But it’s diffing on civet rn and until we can fix that we can’t move forward

**Comment by GiudGiud** [comment]
You could regold with your version and update the SAM submodule in the app folder of the VTB

**Comment by GiudGiud** [comment]
let me try that real quick

**Comment by hapfang** [comment]
Thanks @GiudGiud, I am not exactly sure how I can update the SAM submodule in the app folder of VTB. It shows up as an empty folder in my local repo. What are the two SAM versions civet is using to see the result difference? Thanks.

**Comment by GiudGiud** [comment]
To get a full folder you ll need to git submodule init apps/SAM then git sudmodule update

Let s see if what I did works

**Comment by GiudGiud** [comment]
We still have diffs, on the first time step
https://civet.inl.gov/job/1187377/

**Comment by hapfang** [comment]
@GiudGiud Thanks for the try. What is the SAM version you used to generate the gold reference? And what is the SAM version used by civet? I will report the issue to the SAM team with the input file.

**Comment by GiudGiud** [comment]
They are the same version, I downloaded SAM on my machine from the submodule update included in the PR. 

SAM Version: derived from v0.9.5.1, git commit 50f9d9c7 on 2022-09-21

**Comment by hapfang** [comment]
OK, so does it mean that this specific SAM version produces different results on your local machine and the civet server? I am thinking how we can reproduce the same issue at the ANL side.

**Comment by GiudGiud** [comment]
You could run the test on your computer and see if it diffs?

**Comment by hapfang** [comment]
Thanks Guillaume. The challenge is that I lack the experience of compiling and running SAM locally. My efforts with SAM reactor model development is always done on ANL internal cluster where the SAM executable is periodically updated by the SAM team. 

Could you please share the data files produced on your local machine and civet server? I can also do a short test with 3 steps to compare with them. Hopefully I can get a better idea how the data files differ this way before requesting the help from SAM developers.

**Comment by GiudGiud** [comment]
I actually cant get the files from the civet server. I can get a similar linux machine but not from the test suite. 

My machine's files are on the PR

**Comment by GiudGiud** [comment]
We could do a quick meeting tomorrow if you want to install SAM btw. I have installed it on a few machines now, I kind of need it for VTB maintenance

**Comment by hapfang** [comment]
I think a quick meeting tomorrow will be very helpful. We can discuss how to proceed with this issue. I don't have a suitable machine to run SAM locally. Could you please email me the csv and/or login file you have on your laptop? Thanks.

**Comment by GiudGiud** [comment]
You can download it here
https://github.com/idaholab/virtual_test_bed/blob/4114775336a99e33e35f54176367506ce6c977ed/msr/msre/reactivity_insertion/gold/msre_pke_ss_csv.csv

I m not sure what you mean by login file. 

Ok I m mostly free except 12-2pm please jsut email me an invite

**Comment by hapfang** [comment]
Sorry, that was a typo, and I mean log file. I took a closer look at the gold csv file, and it was generated with varying time step size. However, in the test run on civet server, a fixed time step size (dt = 1) was used for the first 3 steps. I think it could be the cause for the difference you saw. Could you please make sure the reference run and test run use the same time step sizes? Thanks in advance. Sorry for bothering you continuously about this PR. I appreciate your patience.

**Comment by GiudGiud** [comment]
I dont understand why that would be since I pass the same arguments..
I ll check tomorrow. This should be progress thanks for double checking

**Comment by hapfang** [comment]
You could be right, it does look like the first many steps use the constant dt of 1. I will double check tomorrow. 
``` 
  [./time_stepper]
    type = PiecewiseLinear
    x = '-1000.0   0.0'
    y = '    1.0  10.0'
  [../]

  [./TimeStepper]
    type = FunctionDT
    function = time_stepper
    min_dt = 1e-3
  [../]
```

**Comment by hapfang** [comment]
Hi @GiudGiud , for the comparison between test data and gold data, do we expect an exact match, or there is a tolerance? I rerun the latest msre inputs on ANL cluster, and compared the cvs with the gold, differences are noticed, but they are in the range, 1e-11 to 1e-16.

**Comment by GiudGiud** [comment]
In the tests file we have 1e-5 tolerance set. 
I ll look at this today.

**Comment by GiudGiud** [comment]
I m going to loosen the tolerances a little on the diff and tighten on the solve.
Hopefully this will be enough.

**Comment by hapfang** [comment]
Thanks Guillaume. FYI, I compared the csv data of the first 150 step between the gold and the results from an earlier version SAM on ANL cluster. The maximum difference among all the data is 4.6e-7. min difference is 0.

**Comment by GiudGiud** [comment]
Thanks for checking.
I am using exactly the same version of SAM on both machines, this is not a version issue. 

This seems to be differences in convergence properties between two platforms

**Comment by GiudGiud** [comment]
@hapfang 

This is why the results depend on the computer we run this on. The conditioning of this matrix is not great
```
 0 Nonlinear |R| = 1.313843e+03
      SVD: condition number 1.638370744706e+12, 0 of 4089 singular values are (nearly) zero
      SVD: smallest singular values: 3.590959507520e-09 1.780743538134e-08 3.155277067703e-08 5.524547620115e-08 8.482824815616e-08
      SVD: largest singular values : 5.838838684656e+03 5.841991037598e+03 5.865282464742e+03 5.871435848193e+03 5.883323002544e+03
 1 Nonlinear |R| = 7.013315e+02
      SVD: condition number 1.044025430887e+12, 0 of 4089 singular values are (nearly) zero
      SVD: smallest singular values: 5.648760546855e-09 1.185261280924e-08 2.315084003983e-08 4.694500791846e-08 1.180425667857e-07
      SVD: largest singular values : 5.888820862565e+03 5.890093994843e+03 5.893519290118e+03 5.894645664607e+03 5.897449663906e+03
 2 Nonlinear |R| = 6.711227e+02
      SVD: condition number 1.014639334560e+12, 0 of 4089 singular values are (nearly) zero
      SVD: smallest singular values: 5.776606925736e-09 1.226924717340e-08 2.362515795063e-08 4.754866953941e-08 1.064216369797e-07
      SVD: largest singular values : 5.849921214608e+03 5.852603923194e+03 5.857614386518e+03 5.858980934571e+03 5.861172607145e+03
 3 Nonlinear |R| = 5.978653e+02
      SVD: condition number 9.812082648391e+11, 0 of 4089 singular values are (nearly) zero
      SVD: smallest singular values: 5.939981401355e-09 1.281001962377e-08 2.451907942679e-08 4.849236954032e-08 1.082869263425e-07
      SVD: largest singular values : 5.818020765434e+03 5.821035607652e+03 5.821768045217e+03 5.824504108500e+03 5.828358844000e+03
 4 Nonlinear |R| = 5.326339e+02
      SVD: condition number 9.505067416447e+11, 0 of 4089 singular values are (nearly) zero
      SVD: smallest singular values: 6.100464803104e-09 1.335154609755e-08 2.542320369792e-08 4.952149822231e-08 1.134503111689e-07
      SVD: largest singular values : 5.789940675577e+03 5.792801735854e+03 5.793327705991e+03 5.797965770965e+03 5.798532922516e+03
 5 Nonlinear |R| = 4.638587e+02
      SVD: condition number 9.229709096221e+11, 0 of 4089 singular values are (nearly) zero
      SVD: smallest singular values: 6.282601462389e-09 1.392099907935e-08 2.646091050417e-08 5.064108714493e-08 1.171650088475e-07
      SVD: largest singular values : 5.759566348537e+03 5.759755355458e+03 5.762680355851e+03 5.766803537859e+03 5.798658386534e+03
 6 Nonlinear |R| = 3.559983e+02
      SVD: condition number 8.773299870859e+11, 0 of 4089 singular values are (nearly) zero
      SVD: smallest singular values: 6.611138794319e-09 1.484891133906e-08 2.822121757103e-08 5.256945599622e-08 1.197498464684e-07
      SVD: largest singular values : 5.712240766938e+03 5.717701873371e+03 5.722417824135e+03 5.751667402913e+03 5.800150313043e+03
```

**Comment by hapfang** [comment]
I see, any suggestions about how we can improve it? The civet test timed out this time instead.

**Comment by GiudGiud** [comment]
I m fixing the time out. 
With the changes in the tolerances, we can get this to pass on CIVET. So we can now merge this. 

To fix this, you should be preconditioning this matrix by block instead of a global LU

**Comment by hapfang** [comment]
Thanks Guillaume.

**Comment by GiudGiud** [comment]
Nice. Thanks for the contribution!

**Comment by hapfang** [comment]
Thank you for the patience and all the support!

**Pull request merged**